crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,304 @@
+//! Pretty-printable diagnostic reports for custom errors that reference GraphQL documents.
+//!
+//! The [`Diagnostic`] type wraps errors that implement [`ToDiagnostic`].
+use crate::validation::FileId;
+use crate::validation::NodeLocation;
+use crate::SourceFile;
+use crate::SourceMap;
+use ariadne::ColorGenerator;
+use ariadne::ReportKind;
+use std::fmt;
+use std::io;
+use std::ops::Range;
+use std::sync::Arc;
+use std::sync::OnceLock;
+
+#[cfg(doc)]
+use crate::{ExecutableDocument, Schema};
+
+type MappedSpan = (FileId, Range<usize>);
+
+/// Translate a byte-offset location into a char-offset location for use with ariadne.
+fn map_span(sources: &SourceMap, location: NodeLocation) -> Option<MappedSpan> {
+    let source = sources.get(&location.file_id)?;
+    let mapped_source = source.mapped_source();
+    let start = mapped_source.map_index(location.offset());
+    let end = mapped_source.map_index(location.end_offset());
+    Some((location.file_id, start..end))
+}
+
+/// A diagnostic report that can be printed to a CLI with pretty colours and labeled lines of
+/// GraphQL source code.
+///
+/// Custom errors can use this in their `Display` or `Debug` implementations to build a report and
+/// then write it out with [`fmt`].
+///
+/// [`fmt`]: DiagnosticReport::fmt
+pub struct DiagnosticReport {
```

Comment:
Ideas to consider:

* Rename to `CliReport` (since what characterizes Ariadne is monospace text output)
* Replace `SourceMap` parameter in the constructor with `&SourceMap` in `write()` and `fmt`
* Pass color config in the constructor and remove use of `anstream`

Reply:
We need the `SourceMap` to translate locations when you add a label unfortunately.